inductor: fix bf16 legalization issue for fp32 load with to bf16 case

XiaobingSuper opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #103080

Giving following ir:

```
    def body(self, ops):
        get_index = self.get_index('index0')
        index_expr = ops.index_expr(get_index, torch.int32)
        constant = ops.constant(4, torch.int32)
        lt = ops.lt(index_expr, constant)
        masked_subblock1 = self.masked_subblock1(lt, 0.0)
        get_index_1 = self.get_index('index3')
        load = ops.load('arg2_1', get_index_1)
        to_dtype = ops.to_dtype(load, torch.bfloat16)
        where = ops.where(lt, masked_subblock1, to_dtype)
        get_index_2 = self.get_index('index3')
        store = ops.store('buf0', get_index_2, where, None)
        return store
    def masked_subblock2(self, ops):
        get_index = self.get_index('index2')
        load = ops.load('arg1_1', get_index)
        return load
    def masked_subblock1(self, ops):
        get_index = self.get_index('index1')
        index_expr = ops.index_expr(get_index, torch.int32)
        constant = ops.constant(1, torch.int32)
        ge = ops.ge(index_expr, constant)
        get_index_1 = self.get_index('index1')
        index_expr_1 = ops.index_expr(get_index_1, torch.int32)
        constant_1 = ops.constant(3, torch.int32)
        lt = ops.lt(index_expr_1, constant_1)
        and_ = ops.and_(ge, lt)
        masked_subblock2 = self.masked_subblock2(and_, 0.0)
        get_index_2 = self.get_index('index3')
        load = ops.load('arg2_1', get_index_2)
        to_dtype = ops.to_dtype(load, torch.bfloat16)
        where = ops.where(and_, masked_subblock2, to_dtype)
        return where
```

before this PR, the ```masked_subblock2``` will legalize as ```load_bf16+to_fp32```, and the ```masked_subblock2```'s output type is ```fp32```, but for ```load = ops.load('arg2_1', get_index_2), to_dtype = ops.to_dtype(load, torch.bfloat16)```, we didn't convert ```to_bf16``` as ```to_fp32```, which the ```op.where``` has mixed type computation, and will has compiler error: ```error: operands to ?: have different types ‘float’ and ‘c10::BFloat16’```.

This PR will always convert ```to_bf16``` as ```to_fp32``` to fix such an issue.



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @aakhundov

XiaobingSuper committed:
inductor: fix bf16 legalization issue for fp32 load with to bf16 case

[ghstack-poisoned]

XiaobingSuper committed:
Update on "inductor: fix bf16 legalization issue for fp32 load with to bf16 case"


Giving following ir:

```
    def body(self, ops):
        get_index = self.get_index('index0')
        index_expr = ops.index_expr(get_index, torch.int32)
        constant = ops.constant(4, torch.int32)
        lt = ops.lt(index_expr, constant)
        masked_subblock1 = self.masked_subblock1(lt, 0.0)
        get_index_1 = self.get_index('index3')
        load = ops.load('arg2_1', get_index_1)
        to_dtype = ops.to_dtype(load, torch.bfloat16)
        where = ops.where(lt, masked_subblock1, to_dtype)
        get_index_2 = self.get_index('index3')
        store = ops.store('buf0', get_index_2, where, None)
        return store
    def masked_subblock2(self, ops):
        get_index = self.get_index('index2')
        load = ops.load('arg1_1', get_index)
        return load
    def masked_subblock1(self, ops):
        get_index = self.get_index('index1')
        index_expr = ops.index_expr(get_index, torch.int32)
        constant = ops.constant(1, torch.int32)
        ge = ops.ge(index_expr, constant)
        get_index_1 = self.get_index('index1')
        index_expr_1 = ops.index_expr(get_index_1, torch.int32)
        constant_1 = ops.constant(3, torch.int32)
        lt = ops.lt(index_expr_1, constant_1)
        and_ = ops.and_(ge, lt)
        masked_subblock2 = self.masked_subblock2(and_, 0.0)
        get_index_2 = self.get_index('index3')
        load = ops.load('arg2_1', get_index_2)
        to_dtype = ops.to_dtype(load, torch.bfloat16)
        where = ops.where(and_, masked_subblock2, to_dtype)
        return where
```

before this PR, the ```masked_subblock2``` will legalize as ```load_bf16+to_fp32```, and the ```masked_subblock2```'s output type is ```fp32```, but for ```load = ops.load('arg2_1', get_index_2), to_dtype = ops.to_dtype(load, torch.bfloat16)```, we didn't convert ```to_bf16``` as ```to_fp32```, which the ```op.where``` has mixed type computation, and will has compiler error: ```error: operands to ?: have different types ‘float’ and ‘c10::BFloat16’```.

This PR will always convert ```to_bf16``` as ```to_fp32``` to fix such an issue.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225

[ghstack-poisoned]

XiaobingSuper committed:
Update on "inductor: fix bf16 legalization issue for fp32 load with to bf16 case"


Giving following ir:

```
    def body(self, ops):
        get_index = self.get_index('index0')
        index_expr = ops.index_expr(get_index, torch.int32)
        constant = ops.constant(4, torch.int32)
        lt = ops.lt(index_expr, constant)
        masked_subblock1 = self.masked_subblock1(lt, 0.0)
        get_index_1 = self.get_index('index3')
        load = ops.load('arg2_1', get_index_1)
        to_dtype = ops.to_dtype(load, torch.bfloat16)
        where = ops.where(lt, masked_subblock1, to_dtype)
        get_index_2 = self.get_index('index3')
        store = ops.store('buf0', get_index_2, where, None)
        return store
    def masked_subblock2(self, ops):
        get_index = self.get_index('index2')
        load = ops.load('arg1_1', get_index)
        return load
    def masked_subblock1(self, ops):
        get_index = self.get_index('index1')
        index_expr = ops.index_expr(get_index, torch.int32)
        constant = ops.constant(1, torch.int32)
        ge = ops.ge(index_expr, constant)
        get_index_1 = self.get_index('index1')
        index_expr_1 = ops.index_expr(get_index_1, torch.int32)
        constant_1 = ops.constant(3, torch.int32)
        lt = ops.lt(index_expr_1, constant_1)
        and_ = ops.and_(ge, lt)
        masked_subblock2 = self.masked_subblock2(and_, 0.0)
        get_index_2 = self.get_index('index3')
        load = ops.load('arg2_1', get_index_2)
        to_dtype = ops.to_dtype(load, torch.bfloat16)
        where = ops.where(and_, masked_subblock2, to_dtype)
        return where
```

before this PR, the ```masked_subblock2``` will legalize as ```load_bf16+to_fp32```, and the ```masked_subblock2```'s output type is ```fp32```, but for ```load = ops.load('arg2_1', get_index_2), to_dtype = ops.to_dtype(load, torch.bfloat16)```, we didn't convert ```to_bf16``` as ```to_fp32```, which the ```op.where``` has mixed type computation, and will has compiler error: ```error: operands to ?: have different types ‘float’ and ‘c10::BFloat16’```.

This PR will always convert ```to_bf16``` as ```to_fp32``` to fix such an issue.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225

[ghstack-poisoned]

jgong5 commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -2291,12 +2299,13 @@ def is_bf16_mem_copy(node: torch.fx.Node):
                         )
                 elif _node.target == "to_dtype" and _node.args[-1] in [torch.bfloat16]:
                     (ops, x, _) = _node.args
-                    from_load = _node.all_input_nodes[-1].target == "load"
-                    to_store = all(usr.target == "store" for usr in _node.users)
+                    # to_bf16 + store_bf16, after legalization, it should be converted to
+                    # to_bf16 + to_bf16 + store_bf16(before condition), rather than to_fp32 + to_bf16 + store_bf16.
+                    to_store_bf16 = all(is_bf16_store(user) for user in _node.users)
                     # The legalization always loads the BF16 tensor as FP32 for computation and converts
                     # back to BF16 after the computation. Hence, there should be no computation w/ BF16.
                     # Therefore, we update the to_dtype by replacing the bf16 dtype with fp32.
-                    if not (from_load or to_store):
+                    if not to_store_bf16:
```
Can we fix the problem by changing the original logic to `if not to_store`? We only change the dtype to float only when the consumer nodes are compute nodes.

XiaobingSuper committed:
Update on "inductor: fix bf16 legalization issue for fp32 load with to bf16 case"


Giving following ir:

```
    def body(self, ops):
        get_index = self.get_index('index0')
        index_expr = ops.index_expr(get_index, torch.int32)
        constant = ops.constant(4, torch.int32)
        lt = ops.lt(index_expr, constant)
        masked_subblock1 = self.masked_subblock1(lt, 0.0)
        get_index_1 = self.get_index('index3')
        load = ops.load('arg2_1', get_index_1)
        to_dtype = ops.to_dtype(load, torch.bfloat16)
        where = ops.where(lt, masked_subblock1, to_dtype)
        get_index_2 = self.get_index('index3')
        store = ops.store('buf0', get_index_2, where, None)
        return store
    def masked_subblock2(self, ops):
        get_index = self.get_index('index2')
        load = ops.load('arg1_1', get_index)
        return load
    def masked_subblock1(self, ops):
        get_index = self.get_index('index1')
        index_expr = ops.index_expr(get_index, torch.int32)
        constant = ops.constant(1, torch.int32)
        ge = ops.ge(index_expr, constant)
        get_index_1 = self.get_index('index1')
        index_expr_1 = ops.index_expr(get_index_1, torch.int32)
        constant_1 = ops.constant(3, torch.int32)
        lt = ops.lt(index_expr_1, constant_1)
        and_ = ops.and_(ge, lt)
        masked_subblock2 = self.masked_subblock2(and_, 0.0)
        get_index_2 = self.get_index('index3')
        load = ops.load('arg2_1', get_index_2)
        to_dtype = ops.to_dtype(load, torch.bfloat16)
        where = ops.where(and_, masked_subblock2, to_dtype)
        return where
```

before this PR, the ```masked_subblock2``` will legalize as ```load_bf16+to_fp32```, and the ```masked_subblock2```'s output type is ```fp32```, but for ```load = ops.load('arg2_1', get_index_2), to_dtype = ops.to_dtype(load, torch.bfloat16)```, we didn't convert ```to_bf16``` as ```to_fp32```, which the ```op.where``` has mixed type computation, and will has compiler error: ```error: operands to ?: have different types ‘float’ and ‘c10::BFloat16’```.

This PR will always convert ```to_bf16``` as ```to_fp32``` to fix such an issue.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225

[ghstack-poisoned]

XiaobingSuper committed:
Update on "inductor: fix bf16 legalization issue for fp32 load with to bf16 case"


Giving following ir:

```
    def body(self, ops):
        get_index = self.get_index('index0')
        index_expr = ops.index_expr(get_index, torch.int32)
        constant = ops.constant(4, torch.int32)
        lt = ops.lt(index_expr, constant)
        masked_subblock1 = self.masked_subblock1(lt, 0.0)
        get_index_1 = self.get_index('index3')
        load = ops.load('arg2_1', get_index_1)
        to_dtype = ops.to_dtype(load, torch.bfloat16)
        where = ops.where(lt, masked_subblock1, to_dtype)
        get_index_2 = self.get_index('index3')
        store = ops.store('buf0', get_index_2, where, None)
        return store
    def masked_subblock2(self, ops):
        get_index = self.get_index('index2')
        load = ops.load('arg1_1', get_index)
        return load
    def masked_subblock1(self, ops):
        get_index = self.get_index('index1')
        index_expr = ops.index_expr(get_index, torch.int32)
        constant = ops.constant(1, torch.int32)
        ge = ops.ge(index_expr, constant)
        get_index_1 = self.get_index('index1')
        index_expr_1 = ops.index_expr(get_index_1, torch.int32)
        constant_1 = ops.constant(3, torch.int32)
        lt = ops.lt(index_expr_1, constant_1)
        and_ = ops.and_(ge, lt)
        masked_subblock2 = self.masked_subblock2(and_, 0.0)
        get_index_2 = self.get_index('index3')
        load = ops.load('arg2_1', get_index_2)
        to_dtype = ops.to_dtype(load, torch.bfloat16)
        where = ops.where(and_, masked_subblock2, to_dtype)
        return where
```

before this PR, the ```masked_subblock2``` will legalize as ```load_bf16+to_fp32```, and the ```masked_subblock2```'s output type is ```fp32```, but for ```load = ops.load('arg2_1', get_index_2), to_dtype = ops.to_dtype(load, torch.bfloat16)```, we didn't convert ```to_bf16``` as ```to_fp32```, which the ```op.where``` has mixed type computation, and will has compiler error: ```error: operands to ?: have different types ‘float’ and ‘c10::BFloat16’```.

This PR will always convert ```to_bf16``` as ```to_fp32``` to fix such an issue.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225

[ghstack-poisoned]

jgong5 commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -2158,37 +2158,46 @@ def is_bf16_scheduler(self, scheduler_node: SchedulerNode):
+            for _node in sub_graph_nodes:
+                if is_bf16_load_or_constant(_node):
+                    ops = _node.args[0]
+                    with sub_graph.inserting_after(_node):
```
Is it safe to modify the node list `subgraph.nodes` inplace? Should you also record the node list to legalize with and do the graph rewrite after the loop?

XiaobingSuper replied:
Changed to a list to avoid in place change the graph nodes.

XiaobingSuper committed:
Update on "inductor: fix bf16 legalization issue for fp32 load with to bf16 case"


Giving following ir:

```
    def body(self, ops):
        get_index = self.get_index('index0')
        index_expr = ops.index_expr(get_index, torch.int32)
        constant = ops.constant(4, torch.int32)
        lt = ops.lt(index_expr, constant)
        masked_subblock1 = self.masked_subblock1(lt, 0.0)
        get_index_1 = self.get_index('index3')
        load = ops.load('arg2_1', get_index_1)
        to_dtype = ops.to_dtype(load, torch.bfloat16)
        where = ops.where(lt, masked_subblock1, to_dtype)
        get_index_2 = self.get_index('index3')
        store = ops.store('buf0', get_index_2, where, None)
        return store
    def masked_subblock2(self, ops):
        get_index = self.get_index('index2')
        load = ops.load('arg1_1', get_index)
        return load
    def masked_subblock1(self, ops):
        get_index = self.get_index('index1')
        index_expr = ops.index_expr(get_index, torch.int32)
        constant = ops.constant(1, torch.int32)
        ge = ops.ge(index_expr, constant)
        get_index_1 = self.get_index('index1')
        index_expr_1 = ops.index_expr(get_index_1, torch.int32)
        constant_1 = ops.constant(3, torch.int32)
        lt = ops.lt(index_expr_1, constant_1)
        and_ = ops.and_(ge, lt)
        masked_subblock2 = self.masked_subblock2(and_, 0.0)
        get_index_2 = self.get_index('index3')
        load = ops.load('arg2_1', get_index_2)
        to_dtype = ops.to_dtype(load, torch.bfloat16)
        where = ops.where(and_, masked_subblock2, to_dtype)
        return where
```

before this PR, the ```masked_subblock2``` will legalize as ```load_bf16+to_fp32```, and the ```masked_subblock2```'s output type is ```fp32```, but for ```load = ops.load('arg2_1', get_index_2), to_dtype = ops.to_dtype(load, torch.bfloat16)```, we didn't convert ```to_bf16``` as ```to_fp32```, which the ```op.where``` has mixed type computation, and will has compiler error: ```error: operands to ?: have different types ‘float’ and ‘c10::BFloat16’```.

This PR will always convert ```to_bf16``` as ```to_fp32``` to fix such an issue.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225

[ghstack-poisoned]

XiaobingSuper commented:
@pytorchbot merge
